### PR TITLE
Retrofit can return CompletableFutures

### PR DESCRIPTION
**Pull request by j-baker**
Remote calls are currently indistinguishable from local calls, which
makes them hard to use.

Additionally, Retrofit currently sucks to use; the Call api is hard
to use and throws IOExceptions.

This lets it return CompletableFutures, which is better than Feign
and

All of the AsyncCallTracker crap is just there to maintain backcompat;
the current method used to throw a RemoteException is wholly
incompatible with anything async.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/436)

### CODE REVIEW

**Comment by j-baker** [comment]
@markelliot is this the right way to copy Apache 2 code from other places? I remember you knowing how to do this a while back.

Also, I'm sorry about the glaring hack, but I'm not sure there's any other way to do it :(

File: retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/DefaultAsyncCallTracker.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+public final class DefaultAsyncCallTracker implements AsyncCallTracker {
+    private final Set<UUID> asyncCalls = new HashSet<>();
+
+    @Override
+    public <T> void registerAsyncCall(Call<T> call) {
+        asyncCalls.add((UUID) call.request().tag());
+    }
+
+    @Override
+    public boolean isAsyncRequest(Request request) {
```
Comment (uschi2000) [inline retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/DefaultAsyncCallTracker.java]: can we guarantee that this gets called exactly once per registerAsyncCall ?

File: retrofit2-clients/src/test/java/com/palantir/remoting2/retrofit2/Retrofit2ClientApiTest.java [hunk 2]
```diff
@@ -96,6 +106,38 @@ public void testCborRequests() throws IOException, InterruptedException {
...
+        server.enqueue(new MockResponse()
+                .setResponseCode(500)
+                .setHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON)
+                .setBody(ObjectMappers.newClientObjectMapper().writeValueAsString(error)));
+
+        CompletableFuture<String> future = service.makeFutureRequest();
+
+        try {
+            future.join();
+            fail();
```
Comment (uschi2000) [inline retrofit2-clients/src/test/java/com/palantir/remoting2/retrofit2/Retrofit2ClientApiTest.java:134]: use `assertThatThrownBy` (rest of this class uses assertj, i think)

File: retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/AsyncCallTracker.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting2.retrofit2;
+
+import okhttp3.Request;
+import retrofit2.Call;
+
+public interface AsyncCallTracker {
```
Comment (uschi2000) [inline retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/AsyncCallTracker.java]: minimal docs, explain why we need this.

File: retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/AsyncSerializableErrorCallAdapterFactory.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                }
+            };
+
+            call.enqueue(new Callback<R>() {
+                @Override public void onResponse(Call<R> call, Response<R> response) {
+                    if (response.isSuccessful()) {
+                        future.complete(response.body());
+                    } else {
+                        Collection<String> contentTypes = response.raw().headers("Content-Type");
+                        InputStream body = response.errorBody().byteStream();
```
Comment (uschi2000) [inline retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/AsyncSerializableErrorCallAdapterFactory.java:94]: can byteStream be null?

**Review by uschi2000 — CHANGES_REQUESTED** [review]
this doesn't exactly make the retrofit library prettier :(
as an alternative, could we make a static utility that takes a `retrofit2.Call<T>` and turns it into a `CompletableFuture<T>`?

File: retrofit2-clients/src/test/java/com/palantir/remoting2/retrofit2/Retrofit2ClientApiTest.java (hunk 2 above)
Comment (j-baker) [reply retrofit2-clients/src/test/java/com/palantir/remoting2/retrofit2/Retrofit2ClientApiTest.java:134]: you can't assert all that much of the cause, sadly

File: retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/AsyncSerializableErrorCallAdapterFactory.java (hunk 4 above)
Comment (j-baker) [reply retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/AsyncSerializableErrorCallAdapterFactory.java:94]: Don't think so!

**Comment by j-baker** [comment]
Static utility - no, sadly not. The meat of this is in the request adapter, which is essentially something that takes a call and registers a callback; you can't do better because that's the same interface that Call gives you - in fact, this literally wraps Call.

You could wrap it in something like

```
Call<T> call;
CompletableFuture<T> future = new CompletableFuture<>();
try {
    future.complete(call.execute());
} catch (Exception e) {
    future.completeExceptionally(e);
}
```

but this blocks which subverts most of the purpose of doing this.

File: retrofit2-clients/src/test/java/com/palantir/remoting2/retrofit2/Retrofit2ClientApiTest.java (hunk 2 above)
Comment (j-baker) [reply retrofit2-clients/src/test/java/com/palantir/remoting2/retrofit2/Retrofit2ClientApiTest.java:134]: equals, instance, message. It's ugly, can't think of a better way than this.

File: retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/DefaultAsyncCallTracker.java (hunk 1 above)
Comment (j-baker) [reply retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/DefaultAsyncCallTracker.java]: Not without keeping some kind of information about the fact we executed it, which leads to issues.

File: retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/AsyncSerializableErrorCallAdapterFactory.java (hunk 4 above)
Comment (j-baker) [reply retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/AsyncSerializableErrorCallAdapterFactory.java:94]: (errorBody can be, but only if isSuccessful is true)

**Comment by uschi2000** [comment]
for copying code, you can do what the WrappingExecutorService header does, plus add a link:
```
/*
 * Copied from Guava 18.0; original copyright notice below.
 * ... link
 *
 * ... original license ...
 */
```

File: retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/DefaultAsyncCallTracker.java (hunk 1 above)
Comment (uschi2000) [reply retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/DefaultAsyncCallTracker.java]: ok, then i think we should keep these in a cache with timeout.

**Comment by uschi2000** [comment]
i'm not super-sold on the idea of hacking the shit out of retrofit, but i appreciate the future/async is interesting to support. thoughts, @markelliot ?

**Comment by j-baker** [comment]
also - I definitely accept that this doesn't make the retrofit code any better. However, I also think that this entirely transforms the Retrofit interfaces currently writeable;

it turns

```
Call<String> myCall = service.doSomething(args);
try {
    return myCall.execute();
} catch (IOException e) {
    throw new RuntimeException(e);
}
```

into

```
return service.doSomething(args).join();
```

in the most basic case, or in fancier cases, things like

```
return service1.doSomething(args).thenCombine(service2.doSomething(args2), (res1, res2) -> stuff(res1, res2));
```

**Comment by j-baker** [comment]
for example of a real world usage, on the project I've been working on we frequently have to open many streams of data concurrently, and it's really common for the programmer to either just block on them all, or become aware it'd be silly and add an executor service to run their blocking calls (frequently static because no one likes injecting executors all over the place). It means that we have many executors, and frequently our Yourkits just show threads blocking on threads blocking on threads. It'd be really awesome if instead of blocking opening three streams, blocking getting the lengths of each of them, and then blocking reading a small amount of metadata from them, we could just do most of the lifting with the CompletionStage api asynchronously.

**Comment by j-baker** [comment]
@uschi2000 I made it a fair bit less crazy with a touch of mutability. Now it should be basically palatable, no more 'hacking the shit out of retrofit'.

File: retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/Retrofit2ClientBuilder.java [hunk 5]
```diff
@@ -121,6 +123,13 @@ private OkHttpClient createOkHttpClient(String userAgent, List<String> uris) {
         // increase default connection pool from 5 @ 5 minutes to 100 @ 10 minutes
         client.connectionPool(new ConnectionPool(100, 10, TimeUnit.MINUTES));
 
+        Dispatcher dispatcher = new Dispatcher();
```
Comment (uschi2000) [inline retrofit2-clients/src/main/java/com/palantir/remoting2/retrofit2/Retrofit2ClientBuilder.java:126]: sorry, let's keep these in a separate PR.

**Review by uschi2000 — COMMENTED** [review]
one last comment, otherwise i'm fine with this.

**Comment by markelliot** [comment]
Before we merge this (and pick up support load for it), one question we should poke at is whether there's a better way to support James' remoting needs. In particular, would 'dialogue' would be a better fit here?

**Comment by uschi2000** [comment]
It probably would be, but it's not ready.

**Comment by markelliot** [comment]
Ok, it’s not obvious to me this is the right way to proceed.

Why can we not wrap synchronous calls with CompletableFutures external to the library, or instead provide the library function that appropriately turns a Call<T> into a CompletableFuture<T>?

**Comment by uschi2000** [comment]
That was my suggestion above.

**Comment by markelliot** [comment]
Hm, actually, why are you unable to use https://square.github.io/retrofit/2.x/retrofit/retrofit2/Call.html#enqueue-retrofit2.Callback- ?

**Comment by j-baker** [comment]
The RemoteException thing breaks Callback.

**Comment by markelliot** [comment]
I don't understand why, sorry.

**Comment by j-baker** [comment]
I documented why here: https://github.com/palantir/http-remoting/pull/436/files#diff-e75d19c49d5d16f39054b6a769305f37R22 it's the same reason

**Comment by uschi2000** [comment]
Can we re-architect the thing so it throws exceptions in synchronous mode, but plays nicely with the `enqueue()` async mode?

**Comment by uschi2000** [comment]
alternatively, stop throwing exceptions if that's how retrofit is meant to be used.

**Comment by j-baker** [comment]
So in terms of wrapping the synchronous calls external to the library - explicitly I'm trying to achieve more parallelism because there are plenty of places where we do 10 remote calls in a row before being able to consume any data - so if I were to do that I'd have to create my own thread pools and pass them in to each usage site. At the moment, the code I'm looking at has a bunch of static executor services, and I'd rather use the retrofit built in thread pool for this.

I'm hooking this in in the way Retrofit hopes people do this - see https://github.com/square/retrofit/blob/master/retrofit-adapters/java8/src/main/java/retrofit2/adapter/java8/Java8CallAdapterFactory.java - it's the idiomatic way to do it, with a pretty straightforward hoop to jump through http-remoting changes.

In terms of just using Calls and callbacks, the call interface is pretty hard to consume for the user (for example, a pain to mock, since you need to create a Response object which is final) and in a remoting world probably exposes too much to the user.

**Comment by j-baker** [comment]
Stopping throwing exceptions - fine to do that, but this is rather little additional code to avoid the breaking change for consumers (and work required to communicate it) - 25 lines of code.

**Comment by uschi2000** [comment]
Spoke offline; I'm fine taking this. The CompletableFuture style is part of the standard retrofit game ( https://github.com/square/retrofit/blob/master/retrofit-adapters/java8/src/main/java/retrofit2/adapter/java8/Java8CallAdapterFactory.java ) and I agree they are more usable than Calls.

**Review by uschi2000 — APPROVED** [review]

**Comment by markelliot** [comment]
Ok

**Pull request merged**